### PR TITLE
 isolate creation of cross compiler with docker

### LINKED ISSUE #5: Dockerize the process
Hi,

thx a lot for documenting the whole cross compilation process here! However, I'd like to isolate this process and run it inside a docker container in order for my workstation to stay clean and be able to remove all the env vars and configurations one has to make.

I thought about creating a Dockerfile which:
- clones the rust & pi-tools git repos inside the docker image
- sets all needed configs and env vars as specified in the README.md
- compiles the rust cross compiler
- cross compiles a given rust project, also considering optional given binary dependencies 

When running the container with the created image the only things to do are:
- mount a host directory into the container (the individual rust project you want to cross compile)
- optional: mount another host directory into the container including all binary dependencies as .deb files you need (e.g. openssl)
- hand over the cargo cmd (e.g. build --release)
- optional: set an env variable which contains a git commit id/branch/tag for the rust repo to checkout before compiling (default: stable)

**Example usage:**

```
docker run \
    --volume <my rust project directory>:/home/cross/project  \
    --volume <.deb dependency dir>:/home/cross/deps \ # optional
    --env RUST_VERSION=<git reference e.g.: "nightly"> \ # optional
    rust-pi-cross \ # name of created docker image
    release --build
```

What do u think about it? In the next couple of days i want to implement this idea for my own project and if u r interested i can make a pull request... 

Any comments/improvement suggestions welcome!

### PR DESCRIPTION
**Pull request by junactl**
This pull request contains the following changes and thus closes #5:
- implemented a `Dockerfile` which is able to build a `Docker image` which contains the cross compiler (Rust version can specified)
- documented motivation and usage of building the `Docker image` and cross compiling a rust project by running a `Docker container` from that previously built `Docker image`
- split the initial `README.md` into `MANUAL.md` and `DOCKER.md`

### CODE REVIEW

**Comment by Ogeon** [comment]
Nice! I have just mostly skimmed through it (I'll look closer later), but I would like to comment on a few things straight away:
- It looks like the formating of the list is incorrect (press "View" in "File changes"), unless it should be baked into the paragraph.
- Could you mention `DOCKER.md` in `README.md` and maybe even link to it? That would make it easier to discover.
- It's not obvious what the dependency directory is, so it would be nice if it was further explained. Also, would it be possible to make it optional? I think that the more we can slim it down, the better.
- It's also not obvious that the `docker` directory in `cd docker` refers to the directory in this repo. Could you include a `git clone` command or mention that it should be downloaded from here? It's just for the sake of clarity.

It looks good otherwise, as far as I can tell at the moment.

**Comment by junactl** [comment]
Thanks for your feedback. All of your points should be fixed now. I also put the documentation of the manual process into a seperate markdown file just like DOCKER.md and linked to both of them from README.md.

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+ENV TOOLCHAIN_64=$HOME/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin \
+TOOLCHAIN_32=$HOME/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian/bin
+
+# install pi tools
+RUN git clone $URL_GIT_PI_TOOLS $HOME/pi-tools; \
+cd $HOME/pi-tools; \
+git reset --hard $PI_TOOLS_GIT_REF;
+COPY bin/gcc-sysroot $HOME/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin/gcc-sysroot
+COPY bin/gcc-sysroot $HOME/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian/bin/gcc-sysroot
```
Comment (Ogeon) [inline docker/Dockerfile]: I suppose copying from both directories is intentional, right?

File: DOCKER.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# Cross compiling with `Docker`
+The manual process kind of pollutes your workstation by setting lots of environment variables and writing to various config files. Thus it can be difficult to remove or upgrade these files and settings or even repeat that process for different versions of rust on the same machine.
```
Comment (Ogeon) [inline DOCKER.md]: This is a bit exaggerated, to be honest. The "unclean" part would the `gcc-sysroot` script + the config files. The `cross` scripts are meant to take care of and contain the environment variables, to leave the system untouched. So, the parts about various config files and difficulty to remove, upgrade and change version are at least correct.

What I'm trying to say is that there is no real reason to paint the manual method as worse than it is.

**Comment by Ogeon** [comment]
I have taken a closer look and the changes are very good. There is only the concern about the motivation for using Docker, which is a small thing. The mere simplicity of the process is already a big motivator, so that paragraph can be toned down a bit.

The Docker file and the additional scripts are straight forward and looks good. I'm ready to merge this after the motivation thing has been sorted out. It would also be nice if you could edit the description of this pull request to summarize the changes and refer to #5. Preferrably as "Closes #5".

**Comment by Ogeon** [comment]
The new motivation is much better. I like it. Now it's just the PR description, and we are good to go.

File: DOCKER.md (hunk 2 above)
Comment (junactl) [inline DOCKER.md]: Sorry, I read it again and I know what you mean. It was not my intention to comment on the specific implemented process. Just changed this part of the manual accordingly in the latest commit.

Comment (Ogeon) [inline DOCKER.md]: It's alright. That's why it's nice to have someone reviewing the changes :smile: Both methods are good for different reasons, as you wrote in the readme, which is why I prefer to keep things somewhat neutral.

**Comment by Ogeon** [comment]
What I meant was to edit the description (or initial comment) and not the title. I think it's nice to have a short summary at the top of the thread, in case one would like to go back to it again. Nothing fancy, just so it's easy to get an overview. The previous title was alright as it was.

**Comment by junactl** [comment]
Ah, I got it now :)

**Comment by Ogeon** [comment]
Very nice! Thank you so much for the help!

**Pull request merged**